ext/mysqlnd/mysqlnd_auth.c

```diff
@@ -901,6 +1004,73 @@ void php_mysqlnd_scramble_sha2(zend_uchar * const buffer, const zend_uchar * con
 }
 /* }}} */
 
+#ifndef PHP_WIN32
+
+/* {{{ mysqlnd_caching_sha2_public_encrypt */
+static size_t
+mysqlnd_caching_sha2_public_encrypt(MYSQLND_CONN_DATA * conn, mysqlnd_rsa_t server_public_key, size_t passwd_len, unsigned char **crypted, char *xor_str)
+{
+	int server_public_key_len = RSA_size(server_public_key);
+
+	DBG_ENTER("mysqlnd_caching_sha2_public_encrypt");
+	/*
+		Because RSA_PKCS1_OAEP_PADDING is used there is a restriction on the passwd_len.
+		RSA_PKCS1_OAEP_PADDING is recommended for new applications. See more here:
+		http://www.openssl.org/docs/crypto/RSA_public_encrypt.html
+	*/
+	if ((size_t) server_public_key_len - 41 <= passwd_len) {
+		/* password message is to long */
+		SET_CLIENT_ERROR(conn->error_info, CR_UNKNOWN_ERROR, UNKNOWN_SQLSTATE, "password is too long");
+		DBG_ERR("password is too long");
+		DBG_RETURN(0);
+	}
+
+	*crypted = emalloc(server_public_key_len);
```

Comment:
Should this be emalloc? In mysqlnd_sha256_public_encrypt() malloc is used.

Reply:
The return value of `.get_auth_data()` needs to be malloc'd, which is why `mysqlnd_sha256_public_encrypt` mallocs its return value. These `*crypted` don't need to be malloc'd, though (they're efreed() shortly after allocation).